src/webgpu/util/floating_point.ts

```diff
@@ -0,0 +1,3090 @@
+import { assert, unreachable } from '../../common/util/util.js';
+import { Float16Array } from '../../external/petamoriken/float16/float16.js';
+
+import { kValue } from './constants.js';
+import { f32, reinterpretF32AsU32, reinterpretU32AsF32, Scalar } from './conversion.js';
+import {
+  calculatePermutations,
+  cartesianProduct,
+  correctlyRoundedF16,
+  correctlyRoundedF32,
+  flatten2DArray,
+  FlushMode,
+  flushSubnormalNumberF32,
+  isFiniteF16,
+  isFiniteF32,
+  isSubnormalNumberF16,
+  isSubnormalNumberF32,
+  map2DArray,
+  oneULPF32,
+  unflatten2DArray,
+} from './math.js';
+
+/** Indicate the kind of WGSL floating point numbers being operated on */
+export type FPKind = 'f32';
+
+// Containers
+
+/**
+ * Representation of bounds for an interval as an array with either one or two
+ * elements. Single element indicates that the interval is a single point. For
+ * two elements, the first is the lower bound of the interval and the second is
+ * the upper bound.
+ */
+export type IntervalBounds = [number] | [number, number];
+
+/** Represents a closed interval of floating point numbers */
+export class FPInterval {
+  public readonly kind: FPKind;
+  public readonly begin: number;
+  public readonly end: number;
+  // context is JS private, so that objectEquals does not see it when comparing
+  // FPIntervals, since it may not have been initialized yet, but that should
+  // impact equality.
+  #context?: FPContext;
+
+  /**
+   * Constructor
+   *
+   * `FPContext.toInterval` is the preferred way to create FPIntervals
+   *
+   * @param kind the floating point number type this is an interval for
+   * @param bounds beginning and end of the interval
+   */
+  public constructor(kind: FPKind, ...bounds: IntervalBounds) {
+    this.kind = kind;
+
+    const [begin, end] = bounds.length === 2 ? bounds : [bounds[0], bounds[0]];
+    assert(!Number.isNaN(begin) && !Number.isNaN(end), `bounds need to be non-NaN`);
+    assert(begin <= end, `bounds[0] (${begin}) must be less than or equal to bounds[1]  (${end})`);
+
+    this.begin = begin;
+    this.end = end;
+  }
+
+  /** @returns the floating point context for this interval */
+  public context(): FPContext {
+    // context is not fetched and stored at creation time, because there is a
+    // potential dependency cycle between kFPContext and interval on startup.
+    if (this.#context === undefined) {
+      assert(this.kind !== undefined);
+      this.#context = kFPContext[this.kind];
+    }
+    return this.#context;
+  }
```

Comment:
Why do you need to cache this at all? Could you not just use `kFPContext[this.kind]`, or use a getter?